Redesign StripeError and Error-Mapping logic

MattYang-Stripe opened the pull request:
## Summary

<!-- Simple summary of what was changed. -->
We want to introduce the new StripeError design for our RN SDK users.

## PR Description: Optimize Error Mapping Mechanism

### StripeError Structure Redesign

- **Unified Cross-Platform Error Structure**
  - Redesigned `StripeError` interface from simple `{code, message}` to complete `Error` interface
  - Standardized error structure: `name`, `message`, `code`, `nativeErrorCode`, `metadata`
  - Added optional `paymentIntent` and `setupIntent` fields
  - Added TypeScript type safety mechanisms

- **New Error Handling Utilities**
  - `checkIfObjectIsStripeError()`: Type guard function to validate StripeError objects
  - `createStripeError()`: Unified error creation function with defaults and type checking
  - `convertNativeErrorToStripeError()`: Cross-platform error conversion handling iOS (userInfo) and Android (direct properties) differences

### ErrorCode Mapping Logic Improvements

- **iOS Mapping Implementation**
  - Added complete native ErrorCode to React Native ErrorCode mapping logic
  - Uses exhaustive switch for compile-time safety
  - Added `ErrorInformation` struct for unified error information extraction
  - Refactored `mapToStripeErrorObject()` function, separating error information extraction and metadata construction
  - Added `ErrorConstants.swift` for unified error-related constants

- **Android Mapping Optimization**
  - Maintained `TerminalErrorCode.convertToReactNativeErrorCode()` exhaustive when expression
  - Added `ErrorConstants.kt` for unified error constants management
  - Improved error creation and metadata handling

- **Unified ErrorCode Definition**
  - Added `src/Errors/ErrorCodes.ts` providing 60+ standardized error codes
  - Organized by category: Integration, User, Reader/Hardware, Payment, Network, Offline, etc.
  - Replaced legacy `CommonError` enum

- **CI Automation**
  - Added `scripts/check-rn-error-codes.js` script for mapping completeness validation
  - Automatically detects unhandled error codes and cross-platform mapping consistency

- **Dev and Example Applications Updates**
  - Updated dev-app and example-app
  - Updated error handling code to use new `StripeError` interface
  - Migrated from `CommonError` to `ErrorCode` constants
  - Updated TypeScript types and error handling patterns

### Files Requiring Reviewer Attention

#### Core Logic Changes (Critical Review Required)
- `src/types/StripeError.ts` - New StripeError interface definition
- `src/Errors/StripeErrorHelpers.ts` - Core error handling logic and cross-platform conversion
- `src/Errors/ErrorCodes.ts` - Unified error code definitions
- `src/types/index.ts` - Type export updates, removed legacy StripeError and CommonError
- `src/index.tsx` - Added public exports for error handling functions

#### Platform-Specific Implementation
- `ios/Errors.swift` - New iOS error mapping mechanism with ErrorInformation struct
- `ios/ErrorConstants.swift` - New iOS error constants
- `android/src/main/java/com/stripeterminalreactnative/Errors.kt` - Android error mapping optimization
- `android/src/main/java/com/stripeterminalreactnative/ErrorConstants.kt` - New Android error constants

#### New Test Suites
- `ios/Tests/ErrorsTests.swift` - New iOS error handling test suite
- `android/src/test/java/com/stripeterminalreactnative/ErrorsTest.kt` - New Android error handling test suite
- `src/Errors/__tests__/StripeErrorHelpers.test.ts` - New TypeScript test suite

#### Documentation Updates
- `README.md` - Added comprehensive error handling documentation with migration guide and examples

### Key Benefits

1. **Compile-time Safety**: Exhaustive switch/when expressions ensure new error codes must be explicitly handled
2. **Improved Maintainability**: Unified error constants management, reduced code duplication
3. **Cross-platform Consistency**: Unified error structure and handling logic
4. **Automated Validation**: CI scripts ensure mapping completeness and consistency
5. **Type Safety**: TypeScript interfaces and utilities provide better developer experience

### Breaking Changes

**Major Change**: Complete `StripeError` interface redesign

#### Change Details
- **Structure completely different**: From simple `{code, message}` to complete Error interface
- **Removed CommonError enum**: `Failed`, `Canceled`, `Unknown` no longer exist
- **Changed code type**: From generic `T` to concrete `ErrorCode` constants
- **Added required fields**: `name`, `nativeErrorCode`, `metadata`
- **Added optional fields**: `paymentIntent`, `setupIntent`

#### Migration Guide
Detailed migration guide and usage examples are available in the [Error Handling](#error-handling) section of README.md, including:
- Before/after version comparison
- Complete migration steps
- Practical code examples
- Error handling best practices

This change requires users to update existing error handling code to accommodate the new interface. All necessary migration information is included in the updated README documentation.

## Testing

<!-- Did you test your changes? Ideally you should check both of the following boxes. -->

- [X] I tested this manually
- [X] I added automated tests

## Documentation

Select one:

- [X] I have added relevant documentation for my changes.
- [ ] This PR does not result in any developer-facing changes.

bric-stripe reviewed (commented):
I've just skimmed the swift changes and one thing I didn't see is check for the few error types that include resources and some additional data about the API error. We should ensure that data is made available to RN users as well (apologies if this is handled and I just didn't see it):

- https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPConfirmPaymentIntentError.html
- https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPConfirmSetupIntentError.html
- https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPConfirmRefundError.html

the resource and the other additional info would be good to expose

for iOS these will satisfy the "as NSError" since they're subclasses of NSError. So I believe we'll want extra checks in the confirm responses for if the error is the more specialized type 🤔

MattYang-Stripe committed:
Refactor error handling in StripeTerminal SDK to utilize StripeError type and ErrorCode constants. Update functions to return structured error responses. Introduce new ErrorCodes for better error categorization. Enhance type definitions for improved clarity and maintainability.

MattYang-Stripe committed:
Update Android module. Enhance error handling in StripeTerminal SDK by introducing a structured error mapping function. Update error responses to include detailed metadata for StripeError types. Refactor accessibility mapping to handle error cases more gracefully. Remove unused imports for cleaner code.

MattYang-Stripe committed:
Update iOS module. Refactor error handling in StripeTerminal SDK by introducing a comprehensive RNErrorCode enum for structured error categorization. Update error creation methods to utilize the new enum, enhancing clarity and maintainability. Adjust accessibility mapping to return structured error responses.

MattYang-Stripe committed:
Enhance error handling in Errors.kt by adding detailed metadata to unknown error responses. Include exception class and cause information for improved debugging and clarity in error reporting.

MattYang-Stripe committed:
Add yarn step for RN error code checks and implement check script

- Introduced a new yarn step in the Bitrise workflow to run `check:rn-error-codes`.
- Added a new script `check-rn-error-codes.js` to validate consistency of error codes across TypeScript, Swift, and Kotlin implementations.
- Updated `package.json` to include the new script command.

MattYang-Stripe committed:
Fix the root iOS module

MattYang-Stripe committed:
Align the nonStripeError name

MattYang-Stripe committed:
Update the dev-app

MattYang-Stripe committed:
Optimize the ErrorMapping logic in Android and iOS modules

MattYang-Stripe committed:
Update the root Android module test rule

MattYang-Stripe committed:
Add unit test for root android Errors file

MattYang-Stripe committed:
Update root Android module unit test

MattYang-Stripe committed:
Add unit test to root iOS module for Errors swift file

MattYang-Stripe committed:
Optimize the naming of funcitons

MattYang-Stripe committed:
Add Unit Test fo root RN layer

MattYang-Stripe committed:
Remove useless CancellationException test

MattYang-Stripe committed:
Update the error code mapping logic on iOS side

MattYang-Stripe committed:
Clean root RN code

MattYang-Stripe committed:
Update example-app

MattYang-Stripe committed:
Add Error Handling session to the root README

MattYang-Stripe committed:
Add and update readme of the example app

MattYang-Stripe committed:
Remove unused file

MattYang-Stripe committed:
Update the example-app readme

MattYang-Stripe committed:
Format code style

MattYang-Stripe committed:
Fix the typescript error

MattYang-Stripe committed:
Fix typescript issue in the example-app

MattYang-Stripe committed:
Fix the Podfile.lock file of the example-app

MattYang-Stripe committed:
Fix the test problem

MattYang-Stripe committed:
Try to fix the maestro e2e android test

MattYang-Stripe committed:
Revert the e2e test

MattYang-Stripe committed:
Fix e2e Android test

MattYang-Stripe committed:
Temporarily ignore the setupIntent e2e test case due to an unknown CI problem; have tested the case on the local side

MattYang-Stripe committed:
Extract hardcode strings of Errors.kt to constants

MattYang-Stripe committed:
Update README

MattYang-Stripe committed:
Extract hardcoded strings in Error.swift to constants

MattYang-Stripe committed:
Remove CommonError

MattYang-Stripe committed:
Fix the business logic of mapFromReaderSettings

MattYang-Stripe committed:
Optimize Errors.swift

MattYang-Stripe committed:
Optimize Errors.swift

MattYang-Stripe committed:
Optimize StripeErrorHelpers.ts

MattYang-Stripe committed:
Update example ErrorHandling.md

MattYang-Stripe committed:
Remove redundant files

MattYang-Stripe committed:
Remove unused whitespace

MattYang-Stripe committed:
Remove ErrorCode.UNKNOWN enum and update related codes

MattYang-Stripe committed:
Optimize the check-rn-error-codes script

MattYang-Stripe committed:
Optimize the unit test of ios module

MattYang-Stripe committed:
Remove one redundant line in the ErrorsTest.kt

MattYang-Stripe committed:
Add new error-mapping for the three iOS lower-level errors

MattYang-Stripe committed:
Remove a fragile test case in the iOS module

MattYang-Stripe committed:
Update Android side to handle special error classes, containing PI/SI

MattYang-Stripe committed:
Fix iOS unit test

MattYang-Stripe committed:
Update Android side Error Code

MattYang-Stripe committed:
Update the ios module error code to adapt the new native iOS version

MattYang-Stripe committed:
Update RN snapshot test for the new Error Codes

MattYang-Stripe committed:
Fix iOS refund unit tests

MattYang-Stripe committed:
Update README.md

nazli-stripe commented on README.md:
```diff
@@ -187,6 +191,433 @@ class PaymentScreen extends React.Component {
+
+- Over 60 specific error codes organized by category (Network, Reader, Payment, etc.)
+- Additional context like decline codes, API error details, and debugging information
+- Automatic inclusion of related PaymentIntent or SetupIntent when relevant
```
✂️

nazli-stripe commented on README.md:
````diff
@@ -187,6 +191,433 @@ class PaymentScreen extends React.Component {
+  paymentIntent?: { /* Related PaymentIntent */ },
+  setupIntent?: { /* Related SetupIntent */ }
+}
+```
````
since RN is supposed to abstract this information anyways we might as well skip this in the readme too

MattYang-Stripe committed:
Add the remaining fields of the Android APIError to the RN layer

MattYang-Stripe committed:
Trim README

MattYang-Stripe committed:
Update the style of exporting StripeError

MattYang-Stripe committed:
Update Android ErrorsTest unit test

MattYang-Stripe committed:
Trim README.md

bric-stripe commented on README.md:
````diff
@@ -440,6 +440,59 @@ metadata: {
+}
+```
+
+#### Enhanced Error Context for Confirm Operations (iOS)
````
sorry for the delay getting back to look at this 😞 

This is maybe a higher level concern, but ideally we don't have many platform specifics like this in errors, right? Can the RN layer merge things in to a common result? I think a lot of the data mentioned here should all be available from Android as well 🤔 @sjl-stripe please double check me (tagging you since we've had some internal discussions about api error, etc lately). I think some platform specifics are fine for additional details but the declineCode, paymentIntent/setupIntent, etc should be common

https://stripe.dev/stripe-terminal-android/external/com.stripe.stripeterminal.external.models/-terminal-exception/index.html

TerminalException has a nullable `paymentIntent` and `setupIntent` for cases where the confirm fails on the API. refund is missing 🤔 so there may be a gap there.

it also has ApiError which has the declineCode and other goodies[^1] from the API.

iOS embeds those other goodies in the userInfo dictionary. See list of API error keys [here](https://stripe.dev/stripe-terminal-ios/docs/Errors.html#/c:@SCPErrorKeyStripeAPIRequestId)

[^1]: other goodies like docUrl, message, type

MattYang-Stripe committed:
Standardize error handling with discriminated union types and if (result.error) pattern

MattYang-Stripe committed:
Refactor error handling in Stripe Terminal SDK functions to use a dedicated hasError utility function for improved readability and consistency.

MattYang-Stripe committed:
Enhance error resolution in RNSetupIntentCallback by including UUID in error handling for improved traceability.

MattYang-Stripe committed:
Centralize intent error handling to align with Android

Add utility functions to attach PI/SI to all error responses, eliminating code duplication across 10 methods

MattYang-Stripe committed:
Enhance error handling in StripeTerminalReactNative by passing UUID to resolvePaymentIntentError and resolveSetupIntentError methods, ensuring consistent error responses across multiple functions.

MattYang-Stripe committed:
Refactor StripeError

MattYang-Stripe committed:
Fix snapshot test

MattYang-Stripe committed:
Fix iOS unit test

MattYang-Stripe committed:
Update README.md

MattYang-Stripe committed:
Adding detailed documentation to Android & iOS Errors files for error response creation methods

bric-stripe commented on ios/Errors.swift:
```diff
@@ -1,55 +1,667 @@
+            apiError[ErrorConstants.apiErrorMessageKey] = nsError.localizedDescription
+        }
+
+        apiError[ErrorConstants.apiErrorCodeKey] = String(nsError.code)
```
I don't think this is correct. The nsError.code will always be SCPErrorStripeAPIError for API errors.

https://stripe.dev/stripe-terminal-ios/docs/Enums/SCPError.html#/c:@E@SCPError@SCPErrorStripeAPIError

this should be the value from SCPErrorKeyStripeAPIErrorCode if I'm following correctly.

MattYang-Stripe committed:
Update the iOS mechanism of extracting PI/SI/Refund from StripeNSError. Optimize coding style

MattYang-Stripe committed:
Add the missing changes of previous commit

MattYang-Stripe committed:
Add default values to the Android ApiError fields

MattYang-Stripe commented on ios/ErrorConstants.swift:
```diff
@@ -0,0 +1,79 @@
+    static let scpStripeAPIErrorType = ErrorKey.stripeAPIErrorType.rawValue
+    static let scpStripeAPIDocUrl = ErrorKey.stripeAPIDocUrl.rawValue
+    static let scpStripeAPIErrorParameter = ErrorKey.stripeAPIErrorParameter.rawValue
+    static let scpStripeAPICharge = "SCPErrorKeyStripeAPICharge" // Not yet available in current SDK version
```
thank you!

MattYang-Stripe committed:
Fix iOS sytax error

MattYang-Stripe committed:
Fix iOS Error.swift unit test

bric-stripe commented on ios/Errors.swift:
```diff
@@ -184,6 +184,29 @@ class Errors {
+            result[ErrorConstants.setupIntentKey] = Mappers.mapFromSetupIntent(confirmError.setupIntent, uuid: uuid)
+        } else if let confirmError = nsError as? ConfirmRefundError {
+            result[ErrorConstants.refundKey] = Mappers.mapFromRefund(confirmError.refund)
+        }
```
can this go away and reuse `extractResponseObjectFromConfirmError`? 🤔

sjl-stripe replied:
+1. When `createErrorFromNSError(nsError, uuid)` is called, it seems like we run the extraction logic twice:

First extraction (with empty UUID):

1. Line 198 `createErrorFromNSError(nsError)` (no uuid)
2. Which calls `mapToStripeErrorObject(nsError)` (line 184)
3. Which calls `extractResponseObjectFromConfirmError(nsError, &result)` (line 278)
4. This extracts with `uuid: ""` and adds to the error object itself

Second extraction (with UUID):

5. Lines 200-206 extract again with the provided `uuid` parameter
6. This adds to the top level of the response wrapper

The final structure has response objects in two places:

```swift
{
  error: {
    name: "StripeError",
    paymentIntent: { ..., sdkUuid: "" },  // ← First extraction (empty uuid)
    // ...
  },
  paymentIntent: { ..., sdkUuid: "test-uuid" }  // ← Second extraction (provided uuid)
}
```

Android extracts response objects only once at the top level (outside the error object), which is what we probably want?

```kotlin
// android/src/main/java/com/stripeterminalreactnative/Errors.kt:37-46
internal fun WritableMap.putError(throwable: Throwable, uuid: String? = null): ReadableMap = apply {
    putMap(ErrorConstants.ERROR_KEY, nativeMapOf { putErrorContents(throwable) })
    addResponseObjects(throwable, uuid)  // ← Adds to top level, not inside error
}
```

Questions:

1. Is the double extraction intentional, or should we consolidate to match Android's single extraction?
2. I noticed there aren't unit tests for `createErrorFromNSError(nsError, uuid)`. Would it be helpful to add tests that validate this behavior once we confirm what the expected structure should be?

bric-stripe reviewed (commented):
I'm sure I'm missing something 😅 but am happy with where the iOS side of the changes are. thanks for fighting through this and for all the back and forth and updates to consolidate things. some inline comments/questions but nothing blocking assuming manual testing to make sure things are getting set through as expected is ✅ 

leaving unapproved as I think there are still others giving this a look through

sjl-stripe commented on ios/Errors.swift:
```diff
@@ -1,55 +1,665 @@
+        if errorInfo.isStripeError {
+            extractApiErrorToTopLevel(from: nsError, to: &result)
+            extractResponseObjectFromConfirmError(from: nsError, to: &result)
+        }
```
Do all iOS StripeErrors contain an API Error? Asking because `extractApiErrorToTopLevel` always returns a APIError object. This behavior might be different from Android where APIError object isn't created and returned if the SDK doesn't return one.

bric-stripe replied:
good catch 👍 that's correct, the keys will only be filled in if it was a proper API error so 

> If we can't find a stripeAPIErrorCode, then we won't add the ApiError key to the return structure

will work

sjl-stripe reviewed (commented):
Android mapping looks good to me, thanks for doing this! Left a couple comments about iOS mapping.

MattYang-Stripe committed:
Refactor error handling in iOS: streamline error constants and enhance error response structure. Update error mapping methods to align with new constants and improve test coverage for error creation methods.

MattYang-Stripe commented:
> Hi folks, another question. When we receive issues from the Stripe API, is there (or will be) a way to translate them?

Hi @zoontek, thanks for asking this question. cmiiw @sjl-stripe @bric-stripe, currently, we don't. However, our team is planning to provide support to some critical API error messages.

MattYang-Stripe committed:
Fix iOS ErrorsTests

MattYang-Stripe committed:
Refactor UUID handling in payment and setup intents for iOS and Android. Update mapping logic to ensure correct association of intents with UUIDs, improving clarity and consistency in intent management.

MattYang-Stripe commented on ios/StripeTerminalReactNative.swift:
```diff
@@ -531,11 +531,11 @@ class StripeTerminalReactNative: RCTEventEmitter, DiscoveryDelegate, MobileReade
             return
         }
 
+        let uuid = UUID().uuidString
```
Yes, we have updates in some operations. Could you please help me review the business logic below?

## UUID & Response Object Lifecycle


## 🔄 PaymentIntent Complete Lifecycle

| Function | Phase | UUID Source | Map Operation | Return | Status |
|----------|-------|-------------|---------------|---------|---------|
| **createPaymentIntent** | Generate | New UUID | Store: `paymentIntents[uuid] = pi` | `{paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| **retrievePaymentIntent** | Generate | New UUID | Store: `paymentIntents[uuid] = pi` | `{paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| **collectPaymentMethod** | Extract | From input `sdkUuid` | Update: `paymentIntents[uuid] = pi` | `{paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, paymentIntent?: {..., sdkUuid: uuid}}` | ✅ |
| **confirmPaymentIntent** | Extract | From input `sdkUuid` | Clear: `paymentIntents.clear()` / `= [:]` | `{paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, paymentIntent?: {..., sdkUuid: uuid}}` | ✅ |
| **cancelPaymentIntent** | Extract | From input `sdkUuid` | Remove: `paymentIntents[uuid] = null/nil` | `{paymentIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}}` | ✅ |

---

## 🔄 SetupIntent Complete Lifecycle

| Function | Phase | UUID Source | Map Operation | Return | Status |
|----------|-------|-------------|---------------|---------|---------|
| **createSetupIntent** | Generate | New UUID | Store: `setupIntents[uuid] = si` | `{setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| **retrieveSetupIntent** | Generate | New UUID | Store: `setupIntents[uuid] = si` | `{setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| **collectSetupIntentPaymentMethod** | Extract | From input `sdkUuid` | Update: `setupIntents[uuid] = si` | `{setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, setupIntent?: {..., sdkUuid: uuid}}` | ✅ |
| **confirmSetupIntent** | Extract | From input `sdkUuid` | Clear: `setupIntents.clear()` / `= [:]` | `{setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}, setupIntent?: {..., sdkUuid: uuid}}` | ✅ |
| **cancelSetupIntent** | Extract | From input `sdkUuid` | Remove: `setupIntents[uuid] = null/nil` | `{setupIntent: {..., sdkUuid: uuid}}` | ✅ |
| | Error | Same UUID | N/A | `{error: {...}}` | ✅ |

---

## 🐛 Bugs Fixed in This PR

### Android Issues (2 bugs fixed)

1. **`cancelSetupIntent`** ❌ → ✅
   - **Bug**: Used `setupIntent.id.orEmpty()` instead of `uuid`
   - **Fix**: Use `uuid` consistently for map operations

2. **`confirmSetupIntent`** ❌ → ✅
   - **Bug**: Used `it.id.orEmpty()` instead of `uuid`
   - **Fix**: Use `uuid` consistently: `setupIntents.clear()`

---

## 🎯 User Flow Example

```mermaid
sequenceDiagram
    participant User as RN Layer
    participant SDK as Native SDK
    participant Map as In-Memory Map
    
    User->>SDK: createPaymentIntent()
    SDK->>SDK: uuid = "abc-123"
    SDK->>Map: paymentIntents["abc-123"] = PI
    SDK-->>User: {paymentIntent: {stripeId: "pi_xxx", sdkUuid: "abc-123"}}
    
    User->>SDK: collectPaymentMethod({sdkUuid: "abc-123"})
    SDK->>Map: pi = paymentIntents["abc-123"]
    SDK->>Map: paymentIntents["abc-123"] = updatedPI
    SDK-->>User: {paymentIntent: {stripeId: "pi_xxx", sdkUuid: "abc-123"}}
    
    User->>SDK: confirmPaymentIntent({sdkUuid: "abc-123"})
    SDK->>Map: pi = paymentIntents["abc-123"]
    SDK->>Map: paymentIntents.clear()
    SDK-->>User: {paymentIntent: {stripeId: "pi_xxx", sdkUuid: "abc-123"}}
```

**Key Point**: The same `sdkUuid: "abc-123"` is maintained throughout the entire lifecycle, enabling proper object retrieval and map management.

MattYang-Stripe committed:
Update variable names of setupIntent for clarity

MattYang-Stripe committed:
Streamline variable naming and mapping logic for improved clarity and consistency in resolving payment intents

bric-stripe commented on ios/ErrorConstants.swift:
```diff
@@ -0,0 +1,91 @@
+    static let stripeAPIDocUrl = ErrorKey.stripeAPIDocUrl.rawValue
+    static let stripeAPIErrorParameter = ErrorKey.stripeAPIErrorParameter.rawValue
+    // TODO: Replace with ErrorKey.stripeAPICharge.rawValue when available in Stripe Terminal iOS SDK v5.1+
+    static let stripeAPICharge = "com.stripe-terminal:StripeAPIErrorCharge"
```
```suggestion
    static let stripeAPICharge = "com.stripe-terminal:StripeAPICharge"
```

sorry, I was wrong with the first comment...glad I double checked 😅

MattYang-Stripe replied:
thanks! updated

nazli-stripe commented on dev-app/src/screens/CollectCardPaymentScreen.tsx:
```diff
@@ -267,9 +268,16 @@ export default function CollectCardPaymentScreen() {
       if (!resp.client_secret) {
-        return Promise.resolve({
-          error: { message: 'no client_secret returned!' },
+        const error = createStripeError({
```
why is the `createStripeError` being used by the integration? I would expect the SDK to return the error as a StripeError

MattYang-Stripe replied:
I have updated the code, and the design is now below. I think it is suitable for the SDK users.

**SDK Layer**
- Provides `StripeError` as a standardized error structure with rich information (`code`, `message`, `nativeErrorCode`, `metadata`, optional `paymentIntent`/`setupIntent`/`refund`)
- **Security**: Only the SDK can create `StripeError` instances, ensuring payment data comes from trusted sources

**Application Layer**
- SDK users implement their own error classes for app-specific needs, as they may have different error information and structure requirements
- In `dev-app`, we created `DevAppError` to unify error handling across both SDK errors and app-layer errors

### Implementation

**DevAppError** (`dev-app/src/errors/DevAppError.ts`)
- Holds essential fields: `code`, `nativeErrorCode`, `paymentIntent`, `setupIntent`, `refund`, `context`
- Provides `fromStripeError()` static method to convert `StripeError` to `DevAppError`
- Provides `toJSON()` method for consistent log formatting

**Error Handling Utilities** (`dev-app/src/util/errorHandling.ts`)
- Unified functions for error display and logging: `getErrorMessage()`, `getErrorCode()`, `showErrorAlert()`, `showErrorToast()`, `extractErrorInfo()`

nazli-stripe commented on maestro/app.yml:
```diff
@@ -68,12 +68,12 @@ tags:
+#     env:
+#       discoveryMethod: "bt-scn-btn"
+# - runFlow:
+#    file: e2e/setupIntent.yaml
```
is this broken?

xiaoshen-stripe commented:
> Hey @xiaoshen-stripe It's actually not a PoS, but tap-to-pay. Could be english, french, spanish, italian, dutch, german (as I'm aware of). The use-case is a merchant-facing tap-to-pay reader app.
> 
> Apparently, this PR [will expose the `ApiError.code`](https://github.com/stripe/stripe-terminal-react-native/issues/1045#issuecomment-3455447066), allowing us to translate [the API responses](https://docs.stripe.com/error-codes).
> 
> EDIT: It's [here](https://github.com/stripe/stripe-terminal-react-native/pull/1041/files#diff-a27c1e41d3d4d9353d66cd7e2e91459c0203d5ec806f80b1f0cf3781ccc410c2R147) (and I guess the same property is also exposed on iOS)

Thanks 

> Hey @xiaoshen-stripe It's actually not a PoS, but tap-to-pay. Could be english, french, spanish, italian, dutch, german (as I'm aware of). The use-case is a merchant-facing tap-to-pay reader app.
> 
> Apparently, this PR [will expose the `ApiError.code`](https://github.com/stripe/stripe-terminal-react-native/issues/1045#issuecomment-3455447066), allowing us to translate [the API responses](https://docs.stripe.com/error-codes).
> 
> EDIT: It's [here](https://github.com/stripe/stripe-terminal-react-native/pull/1041/files#diff-a27c1e41d3d4d9353d66cd7e2e91459c0203d5ec806f80b1f0cf3781ccc410c2R147) (and I guess the same property is also exposed on iOS)

Thanks @zoontek  super helpful to know this is for Tap to Pay, and you are open to translating yourself using the error code. To double check, you are most interested in Tap to Pay on Android or iPhone (or both?)?

MattYang-Stripe committed:
Update the charge constant in ios/ErrorConstants.swift

Co-authored-by: Brian Cooke <62762265+bric-stripe@users.noreply.github.com>

MattYang-Stripe committed:
Update stripeAPICharge constant in ErrorsTests for accuracy

MattYang-Stripe committed:
Hide the method for creating a Stripe Error and refactor StripeErrorHelpers

MattYang-Stripe committed:
Introduced a new DevAppError class to handle specific errors in the dev-app layer

MattYang-Stripe committed:
Add ExampleAppError class for custom error handling and update API error throwing in example-app

MattYang-Stripe committed:
Add ErrorHandling file to the dev-app and refactor codes

MattYang-Stripe committed:
Optimize ErrorHandling file in the example-app and optimize the code

MattYang-Stripe committed:
Revert maestro temporary changes

MattYang-Stripe committed:
Fix dev-app linter issues

Committed-By-Agent: cursor

zoontek commented:
@xiaoshen-stripe On both platforms

MattYang-Stripe committed:
Add a detailed comment to the check-rn-error-codes.js file to elaborate on the usage. Extract several functions to make the code self-explanatory.

MattYang-Stripe commented on scripts/check-rn-error-codes.js:
```diff
@@ -0,0 +1,208 @@
+  return hasErrors;
+}
+
+function main() {
```
* Extract several functions to make the code self-explanatory.
* Add a comment on the top to elaborate the purpose.

MattYang-Stripe committed:
Refactor DevAppError class to unify error handling across the dev-app. Update screens to utilize the new error handling methods, ensuring consistent error reporting from Stripe SDK errors.
